### PR TITLE
fix(overrides): hoc pattern implementation

### PR DESCRIPTION
**Pull request by chasestarr**
Adresses: https://github.com/uber/baseweb/issues/3103

#### Description

This PR includes the HOC implementation and unit-tests. It has not been added to any components yet

#### Scope

- [x] Patch: Bug Fix

### CODE REVIEW

File: src/styles/__mocks__/styled.js [hunk 1]
```diff
@@ -84,6 +84,7 @@ export function styled(
           ref={this.props.forwardedRef}
           styled-component="true"
           test-style={this.state.styles}
+          test-style-stringified={JSON.stringify(this.state.styles)}
```
Comment (tajo) [inline src/styles/__mocks__/styled.js:87]: Hm, do we really want to add this?  It creates a lot of noise. At least it should be formatted into multiple lines so future diffs are reviewable.

**Comment by tajo** [comment]
Lgtm but the new style attribute should not end up in snapshots.

File: src/styles/__mocks__/styled.js (hunk 1 above)
Comment (chasestarr) [reply src/styles/__mocks__/styled.js:87]: I can format it, but it with react-testing-lib getting the attr `test-style` will only say `[object: Object]`. Do you have any ideas for how to test?

Comment (chasestarr) [reply src/styles/__mocks__/styled.js:87]: I'm going to replace our mock with https://github.com/styletron/styletron/tree/master/packages/styletron-engine-snapshot

**Comment by chasestarr** [comment]
removed the dup styles

**Pull request merged**